Add Yubikey 2FA for unlocking databases

TheZ3ro opened the pull request:
## Description
This just just https://github.com/keepassx/keepassx/pull/52 rebased against keepassxc. All glory belongs to Kyle Manna
This is a merge in `feature/yubikey` from #119 
@johseg you can add commit by pushing to `feature/yubikey` branch

Things to do:
- [x] Add GUI Signals for letting users know when enter the Yubikey https://github.com/keepassxreboot/keepassxc/pull/119#discussion-diff-90652592R169
- [x] YubiKey appears more than once in the list when unlocking the database and then locking it again
- [x] KeePassXC should remember the check state of the "Challenge Response" checkbox. 
- [x] Replace `Q_FOREACH` with C++11 `for` https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11167761
- [x] Replace `Q_DECL_OVERRIDE` with C++11 `override`
- [x] Fix libXtst https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11279305
- [x] Fix Travis build with Yubico lib
- [x] Hiding interface when plugin is disabled or not loaded (based on #125)

See this PR for more #119 



## Motivation and Context
Adds 2FA with yubikeys to keepassxc
Fixes https://github.com/keepassxreboot/keepassxc/issues/26

## How Has This Been Tested?
I've been using his fork for a while. This here is pretty much untested, I use it since today ...

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? If it apply to your pull request, -->
<!--- replace all the `:negative_squared_cross_mark:` with `:white_check_mark:` -->
<!--- Everybody loves emoji -->
- :negative_squared_cross_mark: Bug fix (non-breaking change which fixes an issue)
- :white_check_mark: New feature (non-breaking change which adds functionality)
- :negative_squared_cross_mark: Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, if it apply to your pull request, -->
<!--- replace all the `:negative_squared_cross_mark:` with `:white_check_mark:`. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- Pull Requests that fail the [REQUIRED] field will likely be sent back for corrections or rejected  -->
- :white_check_mark: I have read the **CONTRIBUTING** document. [REQUIRED]
- :negative_squared_cross_mark: My code follows the code style of this project. (I just used the existing code)
- :white_check_mark: All new and existing tests passed. [REQUIRED]
- :white_check_mark:: My change requires a change to the documentation.
- :negative_squared_cross_mark: I have updated the documentation accordingly.
- :white_check_mark:: I have added tests to cover my changes.

Thynix commented:
I'm glad you managed to rebase it, but I'm not a fan of how all the commits in keepassx/keepassx#52 are squished into one. There was information about the design and implementation of the feature in those messages. Thoughts?

What was your reasoning for rebasing the branch instead of merging the current `develop` branch into it?

TheZ3ro commented:
@Thynix this is a merge from #119 where commits were squished unfortunately

kylemanna commented:
@Thynix good call.  Would be nice to preserve the commit history.  There isn't much of a reason not to other then rushing things, and rushing things with a password manager seems reckless.

TheZ3ro commented:
@kylemanna @Thynix the thing is that previous commit were done on a different repo, and now that we are forked the codebase is different and the old commit can have some conflict.
I will try to re-do the merge

kylemanna commented:
> @kylemanna @Thynix the thing is that previous commit were done on a different repo, and now that we are forked the codebase is different and the old commit can have some conflict.
I will try to re-do the merge

The previous commits were done against the upstream keepassx repository, which should have the same ancestors as all forks.  So, that's a relatively moot point.

The real problem is that the patch set has sat on the sidelines and never got merged. As a result, after a few years I stopped rebasing it myself since it was apparent it would never get merged and I didn't want to half ass rebase it and risk corrupting password files for the users that stumbled upon and use my fork.  So, sadly it has sat frozen in time.

But, with the new found attention for the this pull request, let me see what I can do in the next few days to bring this back up to speed. :)

TheZ3ro commented:
@kylemanna Now KeePassXC repo is detached from KeePassX one, we have added other features and fixed things.

This PR is a rebase with fixed conflict on our develop branch. Yes, the original PR's creator (#119) squished all the commit and ***I agree is bad***.

 I've checked manually all the code from this PR and from yours and (a part from minor changes) it's the same code.

I think the commit history is a small problem and that we can get over it, focusing instead on the `Things to do` in the main comment.

>  - Replacing Q_FOREACH with C++11 foreach https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11167761
> - Add GUI Signals for letting users know when enter the Yubikey https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11171617
> - Fix libXtst https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11279305
> - Fix Travis build with Yubico lib
> - Hiding interface when plugin is disabled or not loaded (based on #125)
> - See this PR for more #119

kylemanna commented:
> I think the commit history is a small problem and that we can get over it, focusing instead on the Things to do in the main comment.

I think we're on different pages and the direction of this project seems much different then one I'm interested in supporting.

TheZ3ro commented:
> I think we're on different pages and the direction of this project seems much different then one I'm interested in supporting.
I've just stated my personal opinion about it. I'm not even the one who squished the commit history so there is no need to be "rude" at me.
I, personally, don't have the time to re-do the work someone already done. (*I agree, squishing commit history is bad*)

I'm trying to understanding your rejection to us.
Is it a problem about code copyright and authoring? If yes there is no problem, before the 2.1.0 release we will insert every contributor in the COPYING file.
Is it a problem about being "transparent"? You can review the commit (https://github.com/keepassxreboot/keepassxc/pull/127/commits/7d95bed7380793e6499a99095b6807e73aaf32b8) or better make a new PR with the commit history from https://github.com/keepassx/keepassx/pull/52 (with conflict fixed)

We are maintaining this tool for hobby, no one pays us, we dedicate most of our free time to this project for the community, we are very active and we listen and reply to every single issue. 
Your comment is very sad and negative and I don't think it reflects the current state of the project.

droidmonkey commented:
@kylemanna, I want to make sure we are working closely with the individuals (such as yourself) who put significant effort into previous submissions to the original KeePassX project and were previously ignored.

It is our intention to preserve the integrity and trustworthiness of the code veing submitted. To that end I would like to know what you do not agree with so that we can make corrections where needed.

Your orignal work must be preserved for both credit and posterity. @TheZ3ro, the commits must be unsquished.

TheZ3ro commented:
Fine by me.
I haven't done anything wrong, I haven't squished the commit, I don't have the time to re-do the merge and fixing all the conflicts :)

You guys can re-do the merge and submit a new PR. I'm closing this

phoerious commented on src/keys/drivers/YubiKey.cpp:
```diff
@@ -0,0 +1,250 @@
+YubiKey* YubiKey::m_instance(Q_NULLPTR);
+
+/**
+ * @brief YubiKey::instance - get instance of singleton
```
Thumbs up for API comments 👍, but they should probably go in the header file.

TheZ3ro commented on src/keys/CompositeKey.cpp:
```diff
@@ -64,10 +67,14 @@ CompositeKey& CompositeKey::operator=(const CompositeKey& key)
     clear();
 
-    for (const Key* subKey : asConst(key.m_keys)) {
+    Q_FOREACH (const Key* subKey, key.m_keys) {
```
Duplicate of the previous PR: https://github.com/keepassxreboot/keepassxc/pull/119#pullrequestreview-11167761

phoerious replied:
Alright. But don't forget to use qAsConst() to avoid detaching and unwanted deep copies.

phoerious commented:
Alright, ladies and gents, I rebased Kyle's branch with all commits intact. It was a bit messy because upstream changes were merged in rather than rebased upon, but I got it working. Please test.

I haven't yet changed any of the things I (or @TheZ3ro) annotated during code review. That still needs to be done. There are also two annoying bugs which should be fixed as well before this can be merged:

- YubiKey appears more than once in the list when unlocking the database and then locking it again (got a really long list after a couple of locks and unlocks). This issue already existed in the original branch by Kyle when I tried it a year or two ago.
- KeePassXC should remember the check state of the "Challenge Response" checkbox. Having to check it manually every time is super annoying.

TheZ3ro commented:
Added the 2 new bug to the TODO list in the first comment

kylemanna commented:
@phoerious I just rebased my branch as well, @ https://github.com/kylemanna/keepassxc/tree/feature/yubikey2 without the last commit fixing a rebase in the UI.

I can send a pull request if you'd like.

kylemanna commented:
re @phoerious 

> YubiKey appears more than once in the list when unlocking the database and then locking it again (got a really long list after a couple of locks and unlocks). This issue already existed in the original branch by Kyle when I tried it a year or two ago.

Yes, there are probably a bunch of issues like this.  Additionally unplugging and re-plugging the Yubikey is a potential headache.  There needs to be an event driven system based on USB events implemented, but it was out of scope for the initial code I did.  The goal was to get it merged and not over complicated.

> KeePassXC should remember the check state of the "Challenge Response" checkbox. Having to check it manually every time is super annoying.

Not sure what the default is, but I certainly wouldn't want keepassx remembering what things are required to unlock my databases.  Are the cached somewhere?  Can they be disabled?

I may have never noticed because all of my databases live offline / cold.

kylemanna commented:
To avoid future duplicated work if I can help, please ping me with @kylemanna, I've missed alot of discussion on this topic just because I didn't know it was happening.

phoerious commented:
@kylemanna Hi, sorry, didn't have time to respond earlier. I assumed you were subscribed to this thread, so I didn't ping you directly. Will do so in the future. Sorry also if that led to duplicated work.
Do you have issues with just working on top of this branch here?

droidmonkey commented:
@kylemanna I added you as a collaborator so you can push changes directly to this feature branch.

kylemanna commented:
Hey @phoerious I was subscribed to this thread, my comment about being tagged was re: #119, I never knew that discussion had existed until I happened to stumble on this.

I don't have any issues working on the [feature/yubikey branch](https://github.com/keepassxreboot/keepassxc/tree/feature/yubikey).  First thing I would do is update the HEAD to my private [feature/yubikey2](https://github.com/kylemanna/keepassxc/tree/feature/yubikey2)  assuming @phoerious doesn't have any outstanding work.  Motivation for this is that it doesn't need commit 4497fd49035f501268e42d271f7b3011707d6c45 because I don't think the issue surfaced in my rebase.

phoerious commented:
I don't have any local changes here. I rebased again against the latest commits right before you wrote your comment. Once more than one of us starts working on this branch, we should stop rebasing, though.

kylemanna commented:
Hey @droidmonkey I just attempted to push to the `feature/yubikey` feature branch and get this error:

    ERROR: Permission to keepassxreboot/keepassxc.git denied to kylemanna.

Is there an issue with permissions?

droidmonkey commented:
you need to accept the invitation to join the feature team

kylemanna commented:
@droidmonkey thanks, was able to push.  The invite got priority filed by my mail filters.  I wish Github provided mail headers that made sorting manageable.

kylemanna commented:
Updates to the TODOs:

 > - [ ] Replace Q_FOREACH with C++11 for #119 (review)

* Fixed in commit ef06165ea2f4998e8c04c6c62b04a2fe504e5656

> - [ ] Replace Q_DECL_OVERRIDE with C++11 override
> - [ ] Fix libXtst #119 (review)

* Does not exist in my rebase, perhaps introduced in another rebase?  If still there, please point them out.

>  Fix Travis build with Yubico lib

* Fixed in commit 05774854efb4e32280aac7d7bcd420779c4ec375

kylemanna commented:
Also, the yubikey tests are somewhat arbitrary on Travis-CI because they are skipped due to missing device. 

Test results in a docker ubuntu:trusty container:

    ********* Start testing of TestYubiKeyChalResp *********
    Config: Using QtTest library 5.2.1, Qt 5.2.1
    PASS   : TestYubiKeyChalResp::initTestCase()
    SKIP   : TestYubiKeyChalResp::detectDevices() Unable to connect to YubiKey
       Loc: [/build/tests/TestYkChallengeResponseKey.cpp(45)]
    SKIP   : TestYubiKeyChalResp::getSerial() Unable to connect to YubiKey
       Loc: [/build/tests/TestYkChallengeResponseKey.cpp(45)]
    SKIP   : TestYubiKeyChalResp::keyGetName() Unable to connect to YubiKey
       Loc: [/build/tests/TestYkChallengeResponseKey.cpp(45)]
    SKIP   : TestYubiKeyChalResp::keyIssueChallenge() Unable to connect to YubiKey
       Loc: [/build/tests/TestYkChallengeResponseKey.cpp(45)]
    SKIP   : TestYubiKeyChalResp::deinit() Unable to connect to YubiKey
       Loc: [/build/tests/TestYkChallengeResponseKey.cpp(45)]
    PASS   : TestYubiKeyChalResp::cleanupTestCase()
    Totals: 2 passed, 0 failed, 5 skipped
    ********* Finished testing of TestYubiKeyChalResp *********

Test results on my native Linux workstation with the YubiKey plugged in:

    ********* Start testing of TestYubiKeyChalResp *********
    Config: Using QtTest library 5.7.1, Qt 5.7.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 6.2.1 20160830)
    PASS   : TestYubiKeyChalResp::initTestCase()
    PASS   : TestYubiKeyChalResp::detectDevices()
    PASS   : TestYubiKeyChalResp::getSerial()
    PASS   : TestYubiKeyChalResp::keyGetName()
    PASS   : TestYubiKeyChalResp::keyIssueChallenge()
    PASS   : TestYubiKeyChalResp::deinit()
    PASS   : TestYubiKeyChalResp::cleanupTestCase()
    Totals: 7 passed, 0 failed, 0 skipped, 0 blacklisted, 2261ms
    ********* Finished testing of TestYubiKeyChalResp *********

Don't think there is anything practical we can do about this though other then use it to only detect library breakages via Travis-CI and providing a manual verification for devs later.

droidmonkey commented:
There would have to be a mock yubikey made to conduct the tests... which could be a lot of work

phoerious commented:
I can't even get the test to pass on my own system, even though the YuiKey is inserted.

kylemanna commented:
Updates to the TODOs:

> - [ ]  YubiKey appears more than once in the list when unlocking the database and then locking it again

Resolved in commit f33cd1541941a088b90f463d457f4a9f86aa712a by clearing the list before calling detection code.

kylemanna commented:
@phoerious Can you confirm your challenge-response configuration is correct?

Quick test using standalone tools for slots 1 and 2:

    ykchalresp  -v -1 $(dd if=/dev/urandom count=1 | sha256sum | head -c 64)
    ykchalresp  -v -2 $(dd if=/dev/urandom count=1 | sha256sum | head -c 64)

Example on my system:

    $ ykchalresp  -v -2 $(dd if=/dev/urandom count=1 | sha256sum | head -c 64)
    1+0 records in
    1+0 records out
    512 bytes copied, 4.1123e-05 s, 12.5 MB/s
    Firmware version 3.4.2
    Sending 64 bytes HMAC challenge to slot 2
    505b7c7ef4ca34bec60907fd22e21a00038263c3

phoerious commented:
Yes, output is correct. I can also use it to unlock a KeePassX database. However, the test fails with:

```
Config: Using QtTest library 5.7.1, Qt 5.7.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 6.3.1 20170109)
PASS   : TestYubiKeyChalResp::initTestCase()
QFATAL : TestYubiKeyChalResp::detectDevices() ASSERT: "Crypto::initalized()" in file /home/janek/keepassxc/src/crypto/Random.cpp, line 95
FAIL!  : TestYubiKeyChalResp::detectDevices() Received a fatal error.
   Loc: [Unknown file(0)]
```

I also tested your dropdown menu fix. It works, but you also need to clear the list in the ChangeMasterKeyWidget. The dropdown there is still showing duplicates.

kylemanna commented:
> I also tested your dropdown menu fix. It works, but you also need to clear the list in the ChangeMasterKeyWidget. The dropdown there is still showing duplicates.

Fixed with a7cf39c7cd07fd27eb3d521403ad92d69ef4cf74

Ran the unit test again with a Debug build instead and I see the same error, I'll dig into it and fix it.  This raises a question (for another issue perhaps) though: Shouldn't the unit tests on the Travis-CI build run on Debug builds as well so that the assertions fire?

phoerious commented:
That is actually a good point. The test passes when I compile the tests in Release mode.
Whether Travis should compile in Release or Debug mode is a good question. I honestly think it should do both.

kylemanna commented:
@phoerious Your test case assertion is fixed with 5a3ed27fedd5668bab0b9ffab7a815d1c112884e . Looks like the assertion did its job.

> Whether Travis should compile in Release or Debug mode is a good question. I honestly think it should do both.

I agree.  Probably useful to make the test matrix test all combos of release and debug in addition to gcc and clang compilers.

phoerious commented:
Cool, good work. Tests pass now in Debug mode as well.

What still needs to be implemented is some sort of notification for users of the YubiKey 4. It doesn't affect me with my Neo, but some users need to press a button for HMAC-SHA1 CR.

And we should also remember whether a YubiKey was used or not, otherwise unlocking my DB will drive me nuts. That should be done in the same way as key files are remembered when the option "Remember last key files" is enabled. Recently used key files are saved in the config under "LastKeyFiles". See DatabaseOpenWidget.cpp:90

EugenMayer commented:
Any way for me to join the squad for testing my Yubikey neo against KeepasXC?

HBlitz55 commented:
I set the yubikey option in the cmake-file, compiled and installed [feature/yubikey](https://github.com/kylemanna/keepassxc/tree/feature/yubikey2) version (the extension is listed under help/about), unfortunately the challenge response field stays  grey when I try to create a new database or open a database (It's the same in KeepassX + Yubikey support if the Yubikey is **removed**). My Yubikey 4 is configured so that it requires a click for hmac (My yubikey works with the challenge respone plugin with Keepass under Windows and also with the challenge response version of Keepassx). The Yubikey manager has no issue to find my Yubikey.

Does some else have the same issue?

HBlitz55 commented:
I tried the same with [feature/yubikey](https://github.com/kylemanna/keepassxc/tree/feature/yubikey) (instead of feature/yubikey2) but the field is still grey and empthy (ie I can't set a tick). In comparision KeepassX + challenge response shows my yubikey (`Yubikey[...] Challenge Response - Slot 2 - Press`) and allows me to set a tick.

dummys commented:
I'm testing your build with yubikey nano,neo and 4.
Come back here to tell you if I found some bug. Thanks for such great features

kylemanna commented:
@dummys Multiple keys are out of the scope of this initial pull request.  Certainly a feature that could be explored for adding later.

@EugenMayer you're welcome to build the `feature/yubikey` branch like a standard development build.

DavidHamburg commented:
Hi, I'm currently testing this feature branch with an yubikey 4. The combo box at 'open database' and 'change master key' is always disabled. I can access the yubikey with several other tools and was able to successfully use the challange response with keepass2 on windows. Do you have any ideas or tips for me?

clng3434 commented:
Have you installed all necessary packages before compiling?
´´sudo apt-get -qq install cmake libmicrohttpd10 libmicrohttpd-dev libxi-dev qtbase5-dev libqt5x11extras5-dev qttools5-dev qttools5-dev-tools libgcrypt20-dev zlib1g-dev libxtst-dev xvfb;  fi´

clng3434 commented:
I took the settings and required packages from https://github.com/keepassxreboot/keepassxc/blob/develop/.travis.yml and I can use the yubikey (I need to touch the yubikey, would be great to have a sign that's says that I have to push it). But what I don't get is how to set the HMAC-Master-Key for an existing Database? (I have a Keepass(kdbx) that was created under Windows with the ChallangeResponePlugin but it seems impossible to set the already used HMAC-Master-Key)

clng3434 commented:
Sorry,
´ sudo apt-get -qq install cmake libmicrohttpd10 libmicrohttpd-dev libxi-dev qtbase5-dev libqt5x11extras5-dev qttools5-dev qttools5-dev-tools libgcrypt20-dev zlib1g-dev libxtst-dev xvfb libyubikey-dev libykpers-1-dev´ from https://github.com/keepassxreboot/keepassxc/blob/feature/yubikey/.travis.yml. You probaly have to add the YubiKey development libraries: **libyubikey-dev** and libykpers-1-dev**

mario433 commented:
DavidHamburg: Have you installed all the necessary libs? `sudo apt-get -qq install cmake libmicrohttpd10 libmicrohttpd-dev libxi-dev qtbase5-dev libqt5x11extras5-dev qttools5-dev qttools5-dev-tools libgcrypt20-dev zlib1g-dev libxtst-dev xvfb libyubikey-dev libykpers-1-dev` esp the Yubikey libraries **libyubikey-dev** and **libykpers-1-dev**?

kylemanna: I can use my yubikey and it's shown when i create a new database (but the software doesn't ask fpr the master key of the yubikey). Additionaly, it seems like i can't open Keepass-Databases (Win) that where created in combination with the Challenge-Response Plugin (The Software doesn't ask for the Master-HMAC-Key).

mario433 commented:
@DavidHamburg : Have you installed all the necessary libs? `sudo apt-get -qq install cmake libmicrohttpd10 libmicrohttpd-dev libxi-dev qtbase5-dev libqt5x11extras5-dev qttools5-dev qttools5-dev-tools libgcrypt20-dev zlib1g-dev libxtst-dev xvfb libyubikey-dev libykpers-1-dev` esp the Yubikey libraries **libyubikey-dev** and **libykpers-1-dev**?

@kylemanna : I can use my yubikey and it's shown when i create a new database (but the software doesn't ask fpr the master key of the yubikey). Additionaly, it seems like i can't open Keepass-Databases (Win) that where created in combination with the Challenge-Response Plugin (The Software doesn't ask for the Master-HMAC-Key).

kylemanna commented:
@mario433 The KeePassX database doesn't know the challenge-response master key (and shouldn't for security reasons!), only the Yubikey knows the master key.  The implementation works by sending random bytes of data to Yubikey and using that response to encrypt the database.  As long as you don't regenerate the master challenge response key on the Yubikey this will always work.  If you want ot use multiple Yubikeys (highly recommended for recovery in the event of a lost or damaged Yubikey you'd need to install the identical master key on the devices manually using the Yubikey personalization tools).

It sounds like you're using the Keepass Database plugin which isn't compatible.  If the Keepass plugin requires the master key to be shared, then it seems that its design is flawed and leaks key data when it isn't necessary.

mario433 commented:
@kylemanna : Thank, you very much for you reply.  
(Just for sake of completeness, it's the KeeChallenge plugin for KeePass2 

>When you set the password on your database, you should select Yubikey challenge-response under key providers and click ok. In the window that comes up, copy and paste **the secret from your Yubikey.** You will be prompted to insert your Yubikey and press the button to verify that you entered the correct secret.
>
>Your secret is used as the key to encrypt the database. In order to avoid storing the secret in plain text, we generate a challenge-response pair ahead of time. The challenge is stored to be issued on the next login and the response is used as an AES256 key to encrypt the secret. All relevant data is stored in a xml file in the same directory as your database.
https://brush701.github.io/keechallenge/)

DavidHamburg commented:
@mario433 Yes, I have installed all libraries and also used the parameter -DWITH_XC_YUBIKEY=ON for cmake. 
```
cmake is already the newest version (3.5.2-2ubuntu1).
libgcrypt20-dev is already the newest version (1.7.2-2ubuntu1).
libxi-dev is already the newest version (2:1.7.6-1).
libxtst-dev is already the newest version (2:1.2.2-1).
qtbase5-dev is already the newest version (5.6.1+dfsg-3ubuntu6~4).
zlib1g-dev is already the newest version (1:1.2.8.dfsg-2ubuntu5).
libmicrohttpd-dev is already the newest version (0.9.44+dfsg-1ubuntu2).
libmicrohttpd10 is already the newest version (0.9.44+dfsg-1ubuntu2).
libqt5x11extras5-dev is already the newest version (5.6.1-2build1~1).
libyubikey-dev is already the newest version (1.13-2).
qttools5-dev is already the newest version (5.6.1-2build1~1).
qttools5-dev-tools is already the newest version (5.6.1-2build1~1).
xvfb is already the newest version (2:1.18.4-1ubuntu6.1).
libykpers-1-dev is already the newest version (1.18.0-1~ppa2~yakkety1).
0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.
✔ ~/repos/keepassxc/build [feature/yubikey|✔]
```

phoerious commented:
I added some commits to streamline the UI, add a few checks and remember whether challenge-response was used for each database (only if the user activated the option to remember key files).
There is also a busy indicator underneath the combo box now which is shown while KeePassXC is polling for YubiKeys.

phoerious commented:
I also addressed my own review nit-picks. 😄

kylemanna commented:
Cool @phoerious, I'm pressed for time these days if you or someone else wants to finish the outstanding issues.  Would love to see this merged!

phoerious commented:
I'm on it. One more to go.

phoerious commented:
So, I think, this branch is basically finished and ready for a new review.

When your YubiKey is configured to require user interaction, you get a message saying that you need to touch the button while the user interface is disabled.
I implemented that by running the challenge in a background thread while keeping the GUI thread busy in an a QEventLoop (which blocks execution, but processes events and avoids spinning). That way the challenge() method does not return before the YubiKey sent its response, but the GUI doesn't block either (and can show a nice animated instruction message).

I also adjusted autoreload so that you don't need to touch the button twice because KeePassXC tried to reload the database due to the delay.

phoerious commented:
Little screenshot to illustrate what it looks like:

![screenshot_20170224_190043](https://cloud.githubusercontent.com/assets/911270/23314783/a48ffe8a-fac3-11e6-90eb-830b280ed24f.png)

phoerious commented:
Hmm. One thing that still needs fixing are the unit tests...

TheZ3ro commented:
Actually in file `src/CMakeLists.txt` line *154* or so, needs to be added the feature description for Yubikey, otherwise it won't be visible on Cmake output.

Another thing, with Yubikey disabled (and `libyubikey-dev libykpers-1-dev` installed) I get this error when compiling the Stub:

```
/opt/git/keepassxc/src/keys/drivers/YubiKeyStub.cpp:54:6: error: prototype for ‘bool YubiKey::getSerial(unsigned int&) const’ does not match any in class ‘YubiKey’
 bool YubiKey::getSerial(unsigned int& serial) const
      ^
In file included from /opt/git/keepassxc/src/keys/drivers/YubiKeyStub.cpp:23:0:
/opt/git/keepassxc/src/keys/drivers/YubiKey.h:73:10: error: candidate is: bool YubiKey::getSerial(unsigned int&)
     bool getSerial(unsigned int& serial);
          ^
/opt/git/keepassxc/src/keys/drivers/YubiKeyStub.cpp:61:26: error: prototype for ‘YubiKey::ChallengeResult YubiKey::challenge(int, bool, const QByteArray&, QByteArray&) const’ does not match any in class ‘YubiKey’
 YubiKey::ChallengeResult YubiKey::challenge(int slot, bool mayBlock, const QByt
                          ^
In file included from /opt/git/keepassxc/src/keys/drivers/YubiKeyStub.cpp:23:0:
/opt/git/keepassxc/src/keys/drivers/YubiKey.h:66:21: error: candidate is: YubiKey::ChallengeResult YubiKey::challenge(int, bool, const QByteArray&, QByteArray&)
     ChallengeResult challenge(int slot, bool mayBlock, const QByteArray& challe
                     ^
```

phoerious commented:
Good find. Fixed it.

rockihack commented:
I didn't look at the implementation, but this is vulnerable to replay attacks, isn't it?
By observing it once you can defeat the purpose of this "second factor" completely?

phoerious commented:
Yes, but so is entering a password or a key file. This is less a second factor in the sense of an individual onetime password, but rather a key strengthening measure. And there is no way around it, since we are not dealing with a dynamic online service, but an offline database file.

The challenge, however, is based on the master seed which is randomly generated every time the database is saved. So to mitigate this issue, one could issue a save every time the db is unlocked. But that has two issues:

- It's not atomic, an attacker could intercept the response, copy the db and then replay the response before KeePassXC finishes its save. 
- it is very annoying for users with non-passive YukiKeys because they have to touch the key once for unlocking and then once more for saving the new challenge.

rockihack commented:
> The challenge, however, is based on the master seed which is randomly generated every time the database is saved.

Thats what I wanted to know. Thanks.

a-dma commented:
Hi, Yubico guy here.

Sorry in advance if this is not the right place (I can open a separate issue), but has anybody thought about incorporating U2F rather than challenge-response as an authentication method? I think it would be a good addition.

Some links on the protocol:
- https://fidoalliance.org/download/
- https://developers.yubico.com/U2F/

Happy to provide more info if needed.

TheZ3ro commented on src/CMakeLists.txt:
```diff
@@ -201,7 +210,8 @@ target_link_libraries(keepassx_core
                       ${GPGERROR_LIBRARIES}
-                      ${ZLIB_LIBRARIES})
+                      ${ZLIB_LIBRARIES}
+                      ${YUBIKEY_LIBRARIES})
```
This should be placed on line 207 after keepasshttp and autotype

phoerious commented:
I don't see a particular advantage other than devices may be cheaper. The challenge would always be the same (unless you save the database which causes the master seed to be regenerated) and dynamic features like signature counters cannot be used, because the response must also always be the same. KeePassXC cannot really verify signatures, because whatever is used as "second factor" must be something that can strengthen the encryption key. That can be a signature of something, but it doesn't matter if this is a valid signature or simply a strong of random characters.

volans- commented:
Another advantage is that the HMAC challenge uses one of the 2 Yubikey slots, while the U2F doesn't.

Probably it should be a matter for a different feature request, but I was also wondering about the possibility to use the PIV module of the Yubikey to _simplify_ the unlock operation like it works for the login on Mac OS Sierra.
It basically uses one of the PIV certificates as another way to unlock the encryption key, protected by the PIV module's PIN. So in the normal day by day usage you don't have to input the full long and strong password but just the PIV PIN if the Yubikey is inserted.

phoerious commented:
Login to an operating system is a different use case than a crypto container. I also have pam_u2f running on my PC, but it works completely differently.

I don't know the U2f protocol in-depth, so I don't know if you can get a deterministic response every time you issue the same challenge (I guess you can if the YubiKey doesn't add a counter), but that would be the number one prerequisite for using U2f with KeePassXC.

a-dma commented:
I don't know the internals of KeePassXC, but why do you say that the challenge should be always the same? Couldn't you store the key handle (or a list of key handles) as part of the database? Pretty much like pam-u2f does.

phoerious commented:
The problem is that KeePassXC isn't an exclusive frontend to the password vault. When you log into an online service or a computer to which you don't have physical access to etc., the only way to get past the login form is by entering the correct credentials. This isn't true for your password safe. Your password safe is only a crypto container which is somewhere on your hard drive or even in a remote cloud folder. Therefore every additional second factor authentication that doesn't directly contribute to the main encryption key is useless. KeePassXC doesn't grant you access without using the correct YubiKey? Well, use a different software to decrypt the password safe.

Another problem I see with U2F is that the secret is hard-programmed into the key by the manufacturer, which means that you wouldn't be able to have a backup key in your locker. With HMAC you can program the same secret into two keys, with U2F you can't. Therefore if you lose access to your YubiKey, you lose access to your database. An online service can associate multiple keys with your account, no problem. A password database can't. The only way to accomplish that is by having multiple versions of the same encrypted master key in the KDBX header, but that would probably make KeePassXC incompatible with any other software that uses the KeePass format.

TheZ3ro commented:
@a-dma @volans- The main thing is that the response is used (with the masterkey) to encrypt the database. This increase the security of the database directly, but the response must be fixed otherwise you won't access the database ( https://www.yubico.com/products/services-software/personalization-tools/challenge-response/ )

a-dma commented:
@volans- I see what the issue is, as I said I'm not familiar with the internals of KeePassXC and the database format. Keeping a list of key handles should not be a big problem, but the non-exclusivity and the fact that U2F would not be used strictly for authentication purposes but more as a key strengthening factor might be.
Maybe there is a way to shoehorn the protocol in and still retain it's nice properties, but it requires some more thinking and might end up be format-incompatible.

kylemanna commented:
@a-dma As I understand it U2F is an online protocol that requires a server to authenticate devices. My original goal with this patch was to enable a key hardening technique (as others have described above) that works *completely* offline.  I use KeePassXC as an offline password manager.  Requiring it to connect to the Internet is a non-starter and a different password manager and service all together.

Furthermore, what do you suggest the attack vector is for someone trying to crack a KeePassXC database?  I'm sure they'd attack the master password or encryption key (they already have the encrypted payload, which is usually protected by U2F with online services), which as I understand it, U2F would do nothing to strengthen.

a-dma commented:
@kylemanna Well, not really. There is a server component but it doesn't have to be online. The protocol is challenge-response based, how you get that challenge is up to you. Typically you would send this over https and have the client side built-in into a browser, but that's just a use case (that's why it's called _Universal_ 2nd Factor).
That being said, while it can work completely offline, the goal of the protocol is purely to provide authentication. What it boils down to is a signature verification with a yay/nay result. That combined with what @phoerious said of KeePassXC not being the only possible entry-point to the database probably makes it a non-starter.

phoerious commented:
Yes, the fact that it relies on signature verification rather than providing something like an "additional password" based on a configurable secret, probably makes it useless for securing your database.
But I still think KeepassHTTP would be an interesting use case. Instead of just clicking OK on the confirmation dialog, you have to touch your YukiKey. It remains to be seen if that would actually improve security or if it falls into the category "if someone can simulate button presses on your PC, you have lost already".

phoerious commented:
I had a look at how the KeePass HOTP plugin works. They use a sidecar file to save the encrypted HMAC secret which can be decrypted using the correct sequence of OTPs. It's basically the same thing that we do here, except that we don't support derived OTPs, but require a hardware token to generate the HMAC directly.

U2F isn't supported by KeePass either, probably for the reasons mentioned above.

TheZ3ro commented:
And....Travis is failing

phoerious commented:
I know stupid compiler flags on this ancient GCC. This comes from DEP/ASLR branch which we couldn't test, because GitHub's Travis integration hung itself up. I'm working on it.

droidmonkey commented on src/gui/ChangeMasterKeyWidget.cpp:
```diff
@@ -130,6 +164,23 @@ void ChangeMasterKeyWidget::generateKey()
+        }
+        bool blocking = i & true;
+        int slot      = i >> 1;
+        auto key      = QSharedPointer<YkChallengeResponseKey>(new YkChallengeResponseKey(slot, blocking));
```
The mental gymnastics required to understand this code is crazy. Either comment on what's going on or rename ```i``` to something more descriptive.

droidmonkey replied:
The mental gymnastics required to understand this code is crazy. Either comment on what's going on or rename ```i``` to something more descriptive.

droidmonkey replied:
The mental gymnastics required to understand this code is crazy. Either comment on what's going on or rename ```i``` to something more descriptive.

droidmonkey replied:
The mental gymnastics required to understand this code is crazy. Either comment on what's going on or rename ```i``` to something more descriptive.

phoerious replied:
I'll add a comment. It's basically just a dirty hack to encode more than one piece of information in a combo box entry.

droidmonkey commented on src/gui/DatabaseOpenWidget.cpp:
```diff
@@ -27,11 +27,19 @@
+    : DialogyWidget(parent),
+      m_ui(new Ui::DatabaseOpenWidget()),
+      m_db(nullptr)
 {
```
This does not follow our style.

phoerious replied:
Is it our style to put the commas at the start of the line? I find that insanely ugly.

droidmonkey replied:
it's ugly but functional, makes it easy to add/remove items at the end of the list

droidmonkey commented on src/gui/DatabaseOpenWidget.cpp:
```diff
@@ -160,15 +188,36 @@ CompositeKey DatabaseOpenWidget::databaseKey()
+        bool blocking = i & true;
+        int slot      = i >> 1;
+        auto key      = QSharedPointer<YkChallengeResponseKey>(new YkChallengeResponseKey(slot, blocking));
+        masterKey.addChallengeResponseKey(key);
```
This code is repeated from ChangeMasterKeyWidget::generateKey(), perhaps this needs to be a function?

phoerious replied:
Since this is in two different classes, it would need to be a public function. It doesn't really belong in either class as a public member (or free function), though.

phoerious commented:
Finally merged this baby. Thanks and congrats to all who were participating. Huge milestone. :-)

pejaab commented:
Is this working on MacOS (El Capitan) as well? No matter what i try, installing from dmg or building myself, i am not able to choose a Yubikey based challenge response. Even though Yubikey is listed as extension. Do you have any insights?

phoerious commented:
I suspect you have the 2.1.4 version. Yubikey support requires 2.2.0 which hasn't been released yet.

pejaab commented:
You are correct. Sorry for not mentioning the version number, it was quite late already. Is there an approximate timeframe, when 2.2.0 is going to be released? Or could I build it myself?

phoerious commented:
You can build the develop branch. We do not have a definite release date for 2.2.0 yet, but it won't be too long.

brenthuisman commented:
I'd like to correct a few misconceptions about U2F that have been uttered here. A good start might be to read https://www.yubico.com/about/background/fido/ for those interested. U2F 2.0 will be supported as part of Windows Hello, a sort of authentication replacement for smartcards.

Briefly, nothing about U2F requires online components, and that is why it can and is used for things like (offline) ssh private key authentication. It's basically an easy to use UX around 2 factor, where the master key never leaves the device or can even be read (without cracking the physical key open I suppose). Also, they're the cheapest 2 factor hardware around, which is another reason I hope it's inclusion into KeePaasXC might be reconsidered. In my view it's a perfect fit, and it solves 2 factor for those that are not familiar with TOTP or (not underservedly so imho) consider it not an ideal UX. It comes closest to that real world analog that is so readily understood: the hardware token is a key and inserting it unlocks your website/program/etc. I don't know if it is any more or less safe than TOTP, but I do know that U2F keys I have been able to 'sell' (cheap+ easy to use UX) where TOTP proved bothersome.

The fact that one U2F key can't be backup up is usually circumvented by adding multiple keys.

phoerious commented:
The problem is that U2F devices maintain a counter, so the same challenge doesn't yield the same response, which we need in order to decrypt the database. This is done to prevent device cloning (see https://developers.yubico.com/U2F/Protocol_details/Overview.html#_4_device_cloning_detection), but makes it useless for our purpose. The usual way to allow OTP-based authentication for offline crypto vaults is to pre-calculate a series of OTPs and use them to encrypt a secret which is then used to decrypt the actual crypto vault. But that has a few disadvantages:

- you need an extra sidecar file for storing those pre-calculated OTPs
- you have a fixed secret, which somehow defeats the purpose of 2FA
- if the sidecar file and your device somehow get out of sync, you are screwed

This, however still assumes a deterministic response. The fact that U2F generates a signature from a dynamic component (i.e., the counter), complicates this a lot. We cannot pre-calculate the signature (since we don't know the device's secret) and we don't even know how the counter will increase. It could increase by one, but it could also increase by a 1000. The only requirement is that it is larger than the last time we saw it.

All that considered, I don't see a way to make U2F work for KeePassXC. SSH authentication and WinLogon are totally different scenarios.

brenthuisman commented:
Hmm, I was not aware of that counter, thanks for the pointers. How is that handled by current implementers of U2F, because the key can be used with multiple services simultaneously.

Personally, unlocking my keepass db is a very similar scenario to unlocking my computer/ssh-agent: I want access to something secured.

phoerious commented:
The device generates a signature using public key cryptography and returns both the counter and the signature. An online service can easily verify the signature that way, but we cannot use it as an auxiliary encryption key.

brenthuisman commented:
You mean it can easily verify that because it can easily store the previous counter number?

Online services don't share their counter, yet may share the public key. In that sense, online or offline does not seem to make the difference.

phoerious commented:
A service can store the previous counter (or just ignore it, doesn't matter here), but more importantly, it can verify the signature using the public key. But signature verification isn't the same thing as using the signature to strengthen your encryption key. Both A and B (with A != B) may be valid signatures for a message C, but to strengthen our encryption key, we need need A and B to always be the same if C stays the same, because in the end we need a fixed byte sequence that goes into our encryption key. That's were online and offline differs.

brenthuisman commented:
I think it's dawning on me: in fact it is where authentication and encryption differs. Right? Keepass not only needs authentication, but also a key to encrypt, or some combo, and U2F does not (seem to) provide that, since it's different every time. Useful for authentication (which is why PAM and Hello can use it), encryption not so much.

phoerious commented:
Yes. The counter basically modifies the challenge (i.e., the message). You send a message M to the device and it generates a signature sig(M, c), where c is the counter and then returns both the signature and the counter. So the service can now append M and c and check if the signature is valid for (M, c) and then let you pass (authentication). But a password vault does not have such an authentication scheme. All it has is a fixed encryption key, so we cannot respond to the counter. If we send a challenge M and get a signature for (M, c), which will always be different as c is incremented, we cannot use it to decrypt the database.

brenthuisman commented:
Thanks for helping me understand! Keep up the good work!

tristan-k commented:
I'm running MacOS X `10.11.6` and KeePassXC is unable to detect my YubiKey NEO.

```
$ ioreg -p IOUSB -l -w 0 -x | grep Yubikey -A10
    | +-o Yubikey NEO OTP+CCID@14120000  <class AppleUSBDevice, id 0x1000008a8, registered, matched, active, busy 0 (35 ms), retain 19>
    |     {
    |       "sessionID" = 0x1b6d076c52e
    |       "iManufacturer" = 0x1
    |       "bNumConfigurations" = 0x1
    |       "idProduct" = 0x111
    |       "bcdDevice" = 0x340
    |       "Bus Power Available" = 0x1f4
    |       "USB Address" = 0x9
    |       "bMaxPacketSize0" = 0x40
    |       "iProduct" = 0x2
--
    |       "USB Product Name" = "Yubikey NEO OTP+CCID"
    |       "PortNum" = 0x2
    |       "non-removable" = "no"
    |       "IOCFPlugInTypes" = {"9dc7b780-9ec0-11d4-a54f-000a27052861"="IOUSBFamily.kext/Contents/PlugIns/IOUSBLib.bundle"}
    |       "bDeviceProtocol" = 0x0
    |       "IOUserClientClass" = "IOUSBDeviceUserClientV2"
    |       "IOPowerManagement" = {"DevicePowerState"=0x0,"CurrentPowerState"=0x4,"CapabilityFlags"=0x8000,"MaxPowerState"=0x4,"DriverPowerState"=0x4}
    |       "Device Speed" = 0x1
    |       "USB Vendor Name" = "Yubico"
    |       "idVendor" = 0x1050
    |       "IOGeneralInterest" = "IOCommand is not serializable"
```

![ishot-003](https://user-images.githubusercontent.com/732025/27584670-1d117ce6-5b39-11e7-92f5-a110f86f2bb2.jpg)

phoerious commented:
You need to configure at least one slot to provide HMAC-SHA1.

fluxsauce commented:
@tristan-k @phoerious this should be a FAQ somewhere, I had the same problem and eventually figured it out.

The solution is to use the official - YubiKey Personalization Tool
https://itunes.apple.com/us/app/yubikey-personalization-tool/id638161122?mt=12 - go to Challenge-Response, select Configuration Slot 2, specify the HMAC-SHA1 mode, and Write Configuration.

Shosta commented:
Hi everyone, 

If you need encryption and authentication for the database, why don't you support CCID as well?
For instance, some U2F keys are CCID ready. I think of some FTSafe keys that are cheap (less than 20$).
That would be interesting as well as suppose and it has the encryption and authentication according to Wikipedia (I am not that familiar with that technology).

phoerious commented:
I couldn't find a quick protocol overview for CCID, but it I'll either have the same problems as U2F (please read my other answers for that) or it will serve the same purpose as the HMAC-SHA1 to add entropy to the master key (in which case it would be rather redundant).

hofesh commented:
if I understand this feature correctly, when using a Yubikey, the database uses the master-password plus a random seed, then passed through the yubikey challenge-response, and is encrypted with the Yubikey response. 
So if I have several KeePass files on the same machine, both with the same password and both with the YubiKey step, then even if an attacker compromises my machine while I'm repeatedly using one KeePass file, he won't be able to open the 2nd one, as it's encrypted with a YubiKey response that was created using a different seed. Is that correct?
